### PR TITLE
Preparation for data parallel

### PR DESCRIPTION
**Pull request by lihuoran**
# Description

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [ ] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [ ] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [ ] Mac OS
  - [ ] Linux
- Python version:
  - [ ] 3.6
  - [ ] 3.7
- Key information snapshot(s):

## Needs Follow Up Actions

- [ ] New release package
- [ ] New docker image

## Checklist

- [ ] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/rl_v3/policy_trainer/abs_trainer.py [hunk 1]
```diff
@@ -149,3 +205,22 @@ def set_policy_state_dict(self, policy_state_dict: Dict[str, object]) -> None:
         for policy_name, policy_state in policy_state_dict.items():
             assert policy_name in self._policy_dict
             self._policy_dict[policy_name].set_policy_state(policy_state)
+
+    @abstractmethod
+    def _batch_grad_worker(
```
Comment (buptchan) [inline maro/rl_v3/policy_trainer/abs_trainer.py:210]: `_batch_grad_worker` sounds like a worker prcess/node. 
How about name it as `_get_batch_loss`, and rename `_get_batch_loss` to `get_batch_loss`, since grad_worker will call it as a public method.

File: maro/rl_v3/policy_trainer/abs_trainer.py [hunk 2]
```diff
@@ -19,12 +21,19 @@ def __init__(self, name: str, device: str = None) -> None:
...
         print(f"Creating trainer {self.__class__.__name__} {name} on device {self._device}")
 
+        self._task_queue_client = Optional[TaskQueueClient]
+
     @property
     def name(self) -> str:
         return self._name
 
+    def train_step(self, data_parallel: bool = False) -> None:
+        if data_parallel:
```
Comment (buptchan) [inline maro/rl_v3/policy_trainer/abs_trainer.py:31]: data_parallel -> self._data_parallel

File: maro/rl_v3/policy_trainer/abs_trainer.py [hunk 3]
```diff
@@ -19,12 +21,19 @@ def __init__(self, name: str, device: str = None) -> None:
...
+            assert self._task_queue_client is not None
```
Comment (buptchan) [inline maro/rl_v3/policy_trainer/abs_trainer.py:32]: This can be done during initializing `data_parallel`, instead of checking everytime.
So the `_train_step_impl` can remain as `train_step`.

File: maro/rl_v3/policy_trainer/abs_trainer.py (hunk 1 above)
Comment (lihuoran) [reply maro/rl_v3/policy_trainer/abs_trainer.py]: Renamed to `atomic_get_batch_grad` to explicitly state its functionality: calculating an batch's gradients, in the atomic manner.

File: maro/rl_v3/policy_trainer/abs_trainer.py (hunk 3 above)
Comment (lihuoran) [reply maro/rl_v3/policy_trainer/abs_trainer.py]: I think the assertion in `train_step` is still necessary, because `self._task_queue_client` is not initialized in the `__init__` function. Therefore, `self._task_queue_client` could be `None` even if `self._data_parallel` is `True`.

File: maro/rl_v3/policy_trainer/discrete_maddpg.py [hunk 4]
```diff
@@ -95,122 +95,123 @@ def _register_policies_impl(self, policies: List[RLPolicy]) -> None:
...
-                states=states,  # x
-                actions=actions  # a
-            )  # Q(x, a)
-            critic_loss = self._q_value_loss_func(q_values, target_q_values.detach()) * self._critic_loss_coef
-            critic_losses.append(critic_loss)
-        else:
-            for i in range(self.num_policies):
+                next_actions = [
+                    policy.get_actions_tensor(agent_state)  # a' = miu'(o)
+                    for policy, agent_state in zip(self._target_policies, agent_states)
```
Comment (buptchan) [inline maro/rl_v3/policy_trainer/discrete_maddpg.py:134]: I'm not sure that `agent_states` matches `next_actions`?

File: maro/rl_v3/policy_trainer/distributed_discrete_maddpg.py [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+
+        # Collect next actions
+        next_action_dict: Dict[int, torch.Tensor] = {}
+        for worker in self._workers:
+            next_action_dict.update(worker.get_target_action_dict())
+        next_actions = [next_action_dict[i] for i in range(self.num_policies)]
+
+        # Update critic
+        if self._shared_critic:
+            grads = self._get_batch_grad(
```
Comment (buptchan) [inline maro/rl_v3/policy_trainer/distributed_discrete_maddpg.py:422]: I suggest that leave all the critic & actor update methods in the `MADDPGWorker` to decouple.
If `_shared_critic`, the trainer could maintain a worker locally to update shared critic.

**Review by ysqyang — APPROVED** [review]
